File: src/main/java/com/indiepost/repository/jpa/ActionRepositoryJpa.java
package com.indiepost.repository.jpa;

import com.indiepost.model.analytics.Action;
import com.indiepost.model.analytics.Visitor;
import com.indiepost.repository.ActionRepository;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

@Repository
public class ActionRepositoryJpa implements ActionRepository {

    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public void save(Action action) {
        if (action.getVisitorId() != null) {
            Visitor visitorReference = entityManager.getReference(Visitor.class, action.getVisitorId());
            action.setVisitor(visitorReference);
        }
        entityManager.persist(action);
    }
}

File: src/main/kotlin/com/indiepost/repository/jpa/ActionRepositoryJpa.kt
package com.indiepost.repository.jpa

import com.indiepost.model.analytics.Action
import com.indiepost.model.analytics.Visitor
import com.indiepost.repository.ActionRepository
import org.springframework.stereotype.Repository

import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

@Repository
class ActionRepositoryJpa : ActionRepository {

    @PersistenceContext
    private lateinit var entityManager: EntityManager

    override fun save(action: Action) {
        if (action.visitorId != null) {
            val visitorReference = entityManager.getReference(Visitor::class.java, action.visitorId)
            action.visitor = visitorReference
        }
        entityManager.persist(action)
    }
}
